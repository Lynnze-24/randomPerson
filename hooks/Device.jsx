import { Platform} from "react-native";

export function useDevice() {
  const { OS: platform } = Platform;

 

  const isWeb = platform === "web";
  const isAndroid = platform === "android";
  const isIOS = platform === "ios";

  return {
    isWeb,
    isAndroid,
    isIOS,
    
  };
}