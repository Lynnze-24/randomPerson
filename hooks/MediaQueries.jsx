import React,{ useMemo } from 'react';
import { useWindowDimensions } from 'react-native'

const useMediaQueries = () => {
    const {  width } = useWindowDimensions();
   
    const isLessThanXS = useMemo(() => {
        return width <= 465;
      }, [width]);
    
      const isLessThanSM = useMemo(() => {
        return width <= 550;
      }, [width]);

      const isLessThanXSM = useMemo(() => {
        return width <= 600;
      }, [width]);
    
      const isLessThanMD = useMemo(() => {
        return width <= 700;
      }, [width]);
    
    
      const isLessThanLG = useMemo(() => {
        return width <= 840;
      }, [width]);

      const isLessThanXLG = useMemo(() => {
        return width <= 1000;
      }, [width]);

      const isLessThanXL = useMemo(() => {
        return width <= 1260;
      }, [width]);

      
    

      return {
        width,
        isLessThanXS,
        isLessThanSM,
        isLessThanXSM,
        isLessThanMD,
        isLessThanLG,
        isLessThanXLG,
        isLessThanXL,
      };
}

export default useMediaQueries;
