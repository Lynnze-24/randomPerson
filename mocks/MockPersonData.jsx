export const mockPersonData = [
    {
        name:{
            title:"Mr",
            first:"Aung",
            last:"Linn"
        },
        picture:{
            thumbnail:1
        },
        email:'bbb@gmail.com',
        cell:"999",
        location:{
            city:"Bago",
            country:"Myanmar"
        },
        login:{
            md5:122
        }
        
    },
    {
        name:{
            title:"Mrs",
            first:"Htet",
            last:"Htet"
        },
        picture:{
            thumbnail:1
        },
        email:'aaa@gmail.com',
        cell:"999",
        location:{
            city:"Bago",
            country:"Myanmar"
        },
        login:{
            md5:333
        }
        
    }
]