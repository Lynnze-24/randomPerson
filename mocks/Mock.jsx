import { mockPersonData } from "./MockPersonData"

export const mockVal = {personData:[],
    isLoading:false,
    isError:false,
    refreshPerson:jest.fn(),
    filter:'',
    setFilter:jest.fn(),
    refresh:false,
    isTableView:true,
    switchView:jest.fn()
}




export const mockPersonVal = {...mockVal,
    personData: mockPersonData

}