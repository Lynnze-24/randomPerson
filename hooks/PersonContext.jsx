import React, { createContext, useContext, useState,useEffect} from 'react';
import { mockPersonData } from '../mocks/MockPersonData';



const PersonContext = createContext();


export function usePersonContext(){
    return useContext(PersonContext)
}

export function PersonProvider({children,mock,mockFetch}) {
    const initialData = mockFetch?mockPersonData:[]
    const initialLoading = mockFetch?false:true
   
   const [personData, setPersonData] = useState(initialData); // for person data
    
    const[isLoading,setIsLoading] = useState(initialLoading); // for loading state
    const [isError,setIsError] = useState(false);   // handle fetch error 
    const[refresh,setRefresh] = useState(false);    // for refresh button
    const[filter,setFilter] = useState('');         // setting filter word to filter person data
    
    const[isTableView,setTableView] = useState(true) // switch view

    const switchView = () =>{
        setTableView(!isTableView)
    }

    const fetchPerson = ((mock||mockFetch))?()=>{} 
                            :()=>{
                                fetch('https://randomuser.me/api?results=50')
                                .then(res=> res.json())
                                .then(person=> {
                                        if(isError)setIsError(false)
                                        setPersonData(person.results)
                                        setIsLoading(false)
                                    
                                })
                                .catch(()=>{
                                    setIsLoading(false)
                                    setIsError(true);
                                
                                })
                        }
  

    const refreshPerson = ()=>{
        setIsLoading(true);
        setRefresh(!refresh)
    }
      
    useEffect(() => {
           fetchPerson()
    }, [refresh])
    
    const val = mock? mock : {personData,
                                isLoading,
                                isError,
                                refreshPerson,
                                filter,
                                setFilter,
                                refresh,
                                isTableView,
                                switchView}
    
    return (
        <PersonContext.Provider value={val}>
        {children}
      </PersonContext.Provider>
    );
}
