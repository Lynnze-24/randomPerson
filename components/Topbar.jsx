import React,{useState} from 'react'
import { View, Text,StyleSheet, TouchableHighlight } from 'react-native'
import { TextInput } from 'react-native-paper'
import useMediaQueries from '../hooks/MediaQueries'
import { usePersonContext } from '../hooks/PersonContext'

import cusTheme from '../theme/Theme'
import BarButton from './buttons/BarButton'

const Topbar = ({logout}) => {
    const {isLessThanMD,isLessThanSM} = useMediaQueries()
    const {setFilter,refreshPerson,switchView,isTableView} = usePersonContext()
    const [searchQuery, setSearchQuery] = useState('');    
     const onChangeSearch = query => {
         setSearchQuery(query)
         setFilter(query)
    };      
     
    return (
        <View testID='topBar' style={[styles.container, isLessThanSM && styles.containerSM]}>
           <Text style={[styles.headerText,styles.headerMD]}>Random Person</Text>
           <View style={styles.rightFunc}>
                  <TextInput
                  testID='searchBox'
                    activeUnderlineColor="transparent"
                     style={[styles.icon,styles.inputStyle,
                        isLessThanMD && styles.inputMD,
                        isLessThanSM && styles.inputSM]} 
                     placeholder="Search name"
                     onChangeText={onChangeSearch}
                     value={searchQuery}
                  />
                  <BarButton
                   testID='refreshBtn'
                    icon="refresh"
                    type="MaterialIcons"
                    onPress={refreshPerson}
                  />
                  <BarButton
                    testID='switchBtn'
                    icon={!isTableView?"table":"card-account-details-outline"}
                    type={!isTableView?"AntDesign":"MaterialCommunityIcons"}
                    onPress={switchView}
                  />
               
                <TouchableHighlight testID='logoutButton' onPress={logout} >
                    <Text style={styles.logout}>logout</Text>
                </TouchableHighlight>
           </View>
           
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
      backgroundColor: cusTheme.primary,
      flexDirection:'row',
      height:60,
      paddingHorizontal:50,
      alignItems:'center',
      justifyContent:'space-between'
   },
   containerSM:{
     paddingHorizontal:25
   },
   rightFunc:{
       height:'100%',
        flexDirection:'row',
        alignItems:'center'       
   },
   headerText:{
       color:cusTheme.secondary,
       fontSize:25
   },
   headerMD:{
    fontSize:20
   },
   logout:{
       color:cusTheme.secondary,
       fontSize:18,
       textDecorationLine:'underline'
   },
  
   inputStyle:{
       height:40,
       marginRight:30,
       marginLeft:20,
       width:200,
       borderRadius:5,
   },
   inputMD:{
       width:150
   },
   inputSM:{
    width:120,
    marginRight:10,
    }
})

export default Topbar;
