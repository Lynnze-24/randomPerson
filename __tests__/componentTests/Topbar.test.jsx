import {render, fireEvent } from "@testing-library/react-native";
import Topbar from "../../components/Topbar"
import { PersonProvider } from "../../hooks/PersonContext";
import { mockVal } from "../../mocks/Mock";
import renderer from 'react-test-renderer';



it('matches snapShot', () => {
    const comp = renderer.create( <PersonProvider mock={mockVal}><Topbar/></PersonProvider>).toJSON();
    expect(comp).toMatchSnapshot();
});


it('should render component correctly',()=>{ 
    
    const {getByText,getByTestId}=render(
                     <PersonProvider mock={mockVal}><Topbar/></PersonProvider>
                        );
    getByText('Random Person')
    getByTestId('searchBox')
    getByTestId('refreshBtn')
    getByTestId('switchBtn')
    getByTestId('logoutButton')
     
})


it('should change filter when user type in searchbox',()=>{ 
    
    const {getByTestId}=render(
                     <PersonProvider mock={mockVal}><Topbar/></PersonProvider>
                        );
    const search = getByTestId('searchBox');
   fireEvent.changeText(search,'aa')
   expect(mockVal.setFilter).toHaveBeenCalledWith('aa')
})

it('should call refresh person function when press refresh',()=>{ 
    
    const {getByTestId}=render(
                     <PersonProvider mock={mockVal}><Topbar/></PersonProvider>
                        );
   
   fireEvent.press(getByTestId('refreshBtn'))
   expect(mockVal.refreshPerson).toHaveBeenCalled()
})


it('should call switchView function when press switch icon',()=>{ 
    
    const {getByTestId}=render(
                     <PersonProvider mock={mockVal}><Topbar/></PersonProvider>
                        );
    
   fireEvent.press(getByTestId('switchBtn'))
   expect(mockVal.switchView).toHaveBeenCalled()
})

it('should call logout function when press logout button',()=>{ 
    const logout = jest.fn()
    const {getByTestId}=render(
                     <PersonProvider mock={mockVal}><Topbar logout={logout}/></PersonProvider>
                        );
    
   fireEvent.press(getByTestId('logoutButton'))
   expect(logout).toHaveBeenCalled()
})