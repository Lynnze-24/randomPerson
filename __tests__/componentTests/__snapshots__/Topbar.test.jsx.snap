// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapShot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#120311",
        "flexDirection": "row",
        "height": 60,
        "justifyContent": "space-between",
        "paddingHorizontal": 50,
      },
      false,
    ]
  }
  testID="topBar"
>
  <Text
    style={
      Array [
        Object {
          "color": "#f0edf6",
          "fontSize": 25,
        },
        Object {
          "fontSize": 20,
        },
      ]
    }
  >
    Random Person
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": "100%",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(231, 231, 231)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          Object {
            "borderRadius": 5,
            "marginLeft": 20,
            "marginRight": 30,
            "width": 200,
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.26)",
            "bottom": 0,
            "height": 2,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "transform": Array [
              Object {
                "scaleY": 0.5,
              },
            ],
            "zIndex": 1,
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            Object {
              "minHeight": 40,
            },
          ]
        }
      >
        <TextInput
          editable={true}
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Search name"
          placeholderTextColor="rgba(0, 0, 0, 0.54)"
          selectionColor="transparent"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
              },
              Object {
                "paddingLeft": 12,
                "paddingRight": 12,
              },
              Object {
                "height": 40,
              },
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              false,
              Array [
                Object {},
              ],
            ]
          }
          testID="searchBox"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 15,
              "height": 30,
              "width": 30,
            },
            false,
            Array [
              Object {
                "marginLeft": 0,
                "marginRight": 30,
              },
              false,
            ],
          ],
        ]
      }
      testID="refreshBtn"
    >
      <View>
        <Text />
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 15,
              "height": 30,
              "width": 30,
            },
            false,
            Array [
              Object {
                "marginLeft": 0,
                "marginRight": 30,
              },
              false,
            ],
          ],
        ]
      }
      testID="switchBtn"
    >
      <View>
        <Text />
      </View>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="logoutButton"
    >
      <Text
        style={
          Object {
            "color": "#f0edf6",
            "fontSize": 18,
            "textDecorationLine": "underline",
          }
        }
      >
        logout
      </Text>
    </View>
  </View>
</View>
`;
